src/util/address.rs

```diff
@@ -137,7 +137,143 @@ impl From<bech32::Error> for Error {
     }
 }
 
-/// The different types of addresses.
+/// Detailed address information returned by [`Address::address_info`]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
+// #[non_exhaustive]
+pub enum AddressInfo {
+    /// Witness versions recognized by the current consensus
+    Known(AddressType),
+    /// Future witness versions (>1)
+    Future(WitnessVersion, /** witness program length */ usize),
+    /// Witness version is 0 and is known to consensus, but witness program has non-standard length
+    /// which is not supported by consensus rules and can't be spent.
+    NotSpendableWitnessV0(/** witness program length */ usize),
```

Comment:
As others have mentioned, I find this variant very confusing. There are lots of ways to construct provably unspendable outputs, and this one is particularly obscure. It's also not clear that you could actually produce a usable address for something like this, even if you bech32-encoded it, because parsers tend to have length checks.

Similarly `NonTaprootWitnessV1` should be merged into `Future`.